src/components/themed/Carousel.tsx

```diff
@@ -0,0 +1,202 @@
+import * as React from 'react'
+import { useEffect } from 'react'
+import { LayoutChangeEvent, Pressable, View } from 'react-native'
+import { Gesture, GestureDetector } from 'react-native-gesture-handler'
+import Animated, { runOnJS, SharedValue, useAnimatedStyle, useSharedValue, withSpring, withTiming } from 'react-native-reanimated'
+
+import { useWindowSize } from '../../hooks/useWindowSize'
+import { useState } from '../../types/reactHooks'
+import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
+
+interface Props<T> {
+  items: T[]
+  renderItem: (item: T, index: number) => React.ReactNode
+  keyExtractor?: (item: T, index: number) => string
+  onChangeItem?: (item: T, index: number) => void
+}
+
+interface ItemDisplayProps {
+  children: React.ReactNode
+  currentOffset: SharedValue<number>
+  itemIndex: number
+  onLayout: (event: LayoutChangeEvent) => void
+}
+
+interface PaginationDotProps {
+  currentOffset: SharedValue<number>
+  itemIndex: number
+  onPress?: () => void
+}
+
+export function Carousel<T>(props: Props<T>) {
+  const theme = useTheme()
+  const styles = getStyles(theme)
+  const { items, keyExtractor = defaultKeyExtractor, renderItem, onChangeItem } = props
+
+  const handleChangeItem = (index: number) => {
+    if (onChangeItem != null) {
+      onChangeItem(items[index], index)
+    }
+  }
+
+  const handlePressPaginationDot = (index: number) => {
+    offset.value = withTiming(index)
+    handleChangeItem(index)
+  }
+
+  // Respond to device orientation changes:
+  const { width: windowWidth } = useWindowSize()
+  useEffect(() => {
+    setItemWidth(itemWidth => itemWidth)
+  }, [windowWidth])
+
+  const itemCount = items.length
+  const [itemWidth, setItemWidth] = useState(1) // 1 avoids division by zero
+  const trackWidth = itemWidth * itemCount
+  const handleLayout = (event: LayoutChangeEvent) => {
+    // The check prevents layout jittering
+    if (Math.abs(event.nativeEvent.layout.width - itemWidth) >= 1) {
+      setItemWidth(event.nativeEvent.layout.width)
+    }
+  }
```

Comment:
This whole calculation seems sketchy. If I were doing it, I think I would measure the outer `Animated.View` container, to see how wide / tall he ends up being in the context of the scene. Then, I would set the children's `maxHeight / maxWidth` to match the container, so no matter what aspect ratio the container has, a whole child will always appear. Finally, I would measure where the children actually land in terms of aspect ratio, and use those locations for detent positions.

However, I see that your logic does work, given the band-aids (orientation detection and de-bounce protection). So we can keep the current logic if you prefer, accepting it's limitations.

Reply:
I think for the sake of time, we can keep the impl. I have which does work and is thoroughly dev-tested, and if I have some time during the review time of the sprint, I can try a second attempt before we merge this PR.